sdk/azidentity/client_secret_credential.go

```diff
@@ -54,47 +55,65 @@ func (c *ClientSecretCredential) AuthenticationPolicy(options azcore.Authenticat
 }
 
 type bearerTokenPolicy struct {
-	// take lock when manipulating header/expiresOn fields
-	lock      sync.RWMutex
-	header    string
-	expiresOn time.Time
+	updating  atomic.Int64 // atomically set to 1 to indicate a refresh is in progress
+	header    atomic.String
+	expiresOn atomic.Time
 	creds     azcore.TokenCredential // R/O
 	scopes    []string               // R/O
 }
 
 func newBearerTokenPolicy(creds azcore.TokenCredential, scopes []string) *bearerTokenPolicy {
-	// set the token as expired so first call to Do() refreshes it
-	return &bearerTokenPolicy{creds: creds, scopes: scopes, expiresOn: time.Now().UTC()}
+	bt := bearerTokenPolicy{creds: creds, scopes: scopes}
+	// set initial values to their zero-value
+	bt.header.Store("")
+	bt.expiresOn.Store(time.Time{})
+	return &bt
 }
 
 func (b *bearerTokenPolicy) Do(ctx context.Context, req *azcore.Request) (*azcore.Response, error) {
+	// create a "refresh window" before the token's real expiration date.
+	// this allows callers to continue to use the old token while the
+	// refresh is in progress.
+	const window = 2 * time.Minute
 	var bt string
-	// take read lock and check if the token has expired
-	b.lock.RLock()
+	// check if the token has expired
 	now := time.Now().UTC()
-	if now.Equal(b.expiresOn) || now.After(b.expiresOn) {
-		// token has expired, take the write lock then check again
-		b.lock.RUnlock()
-		// don't defer Unlock(), we want to release it ASAP
-		b.lock.Lock()
-		if now.Equal(b.expiresOn) || now.After(b.expiresOn) {
-			// token has expired, get a new one and update shared state
+	exp := b.expiresOn.Load()
+	if now.After(exp.Add(-window)) {
+		// token has expired, set update flag and begin the refresh.
+		// if no other go routine has initiated a refresh the calling
+		// go routine will do it.
+		if b.updating.CAS(0, 1) {
 			tk, err := b.creds.GetToken(ctx, b.scopes)
 			if err != nil {
-				b.lock.Unlock()
+				// clear updating flag before returning
+				b.updating.Store(0)
 				return nil, err
 			}
-			b.expiresOn = tk.ExpiresOn
-			b.header = "Bearer " + tk.Token
-		} // else { another go routine already refreshed the token }
-		bt = b.header
-		b.lock.Unlock()
+			// we must set header before expiresOn.  this is to prevent a race
+			// when setting header for the very first time.  if expiresOn is set
+			// first it's possible to get an empty header.
+			b.header.Store("Bearer " + tk.Token)
+			b.expiresOn.Store(tk.ExpiresOn)
+			// signal the refresh is complete. this must happen after all shared
+			// state has been updated.
+			b.updating.Store(0)
+		} // else { another go routine is refreshing the token }
+		for {
+			bt = b.header.Load()
+			// for the very first call to Do() the header will be
+			// the empty string.  in this case we need to spin-wait
+			// until header contains a value.  subsequent calls to
+			// Do() will never spin-wait.
+			if bt != "" {
+				break
+			}
+			runtime.Gosched()
```

Comment:
This Gosched bothers me. I'd prefer a time.Sleep() with 100ms or so.
There is another potential problem. If a thread tried to get the token and this errors out, then we might have goroutines spinning forever and none of them ever try to get the token again and then never error out.

So, if bt == "", then these go routine should also spin larger trying to do the CAS to try to get the token.

Reply:
A few other spin-lock implementations I found online use Gosched(), it yields the processor without suspending the calling go routine compared to sleep which will suspend the calling go routine.
You're right about the error case, I was thinking about that this morning.  Ideally we should exit the spin-wait and retry from the beginning (will work on this).